pkg/build/controller/policy/policy.go

```diff
@@ -0,0 +1,116 @@
+	for _, s := range policies {
+		switch buildutil.BuildRunPolicy(build) {
+		case buildapi.BuildRunPolicyParallel:
+			if _, ok := s.(*ParallelPolicy); ok {
+				glog.V(5).Infof("Using %T run policy for build %s/%s", s, build.Namespace, build.Name)
+				return s
+			}
+		case buildapi.BuildRunPolicySerial:
+			if _, ok := s.(*SerialPolicy); ok {
+				glog.V(5).Infof("Using %T run policy for build %s/%s", s, build.Namespace, build.Name)
+				return s
+			}
+		case buildapi.BuildRunPolicySerialLatestOnly:
+			if _, ok := s.(*SerialLatestOnlyPolicy); ok {
+				glog.V(5).Infof("Using %T run policy for build %s/%s", s, build.Namespace, build.Name)
+				return s
+			}
+		}
+	}
+	return nil
+}
+
+// GetNextConfigBuild returns the build that will be executed next for the given
+// build configuration. It also returns the indication whether there are
+// currently running builds, to make sure there is no race-condition between
+// re-listing the builds.
+func GetNextConfigBuild(lister buildclient.BuildLister, namespace, buildConfigName string) (*buildapi.Build, bool, error) {
+	var (
+		nextBuild           *buildapi.Build
+		hasRunningBuilds    bool
+		previousBuildNumber int64
+	)
+	builds, err := buildutil.BuildConfigBuilds(lister, namespace, buildConfigName, func(b buildapi.Build) bool {
+		switch b.Status.Phase {
+		case buildapi.BuildPhasePending, buildapi.BuildPhaseRunning:
+			hasRunningBuilds = true
+			return false
+		}
+		// Only 'new' build can be scheduled to run next
+		return b.Status.Phase == buildapi.BuildPhaseNew
+	})
+	if err != nil {
+		return nil, hasRunningBuilds, err
+	}
+
+	for i, b := range builds.Items {
+		buildNumber, err := buildutil.BuildNumber(&b)
+		if err != nil {
+			return nil, hasRunningBuilds, err
+		}
+		if previousBuildNumber == 0 || buildNumber < previousBuildNumber {
+			nextBuild = &builds.Items[i]
+			previousBuildNumber = buildNumber
+		}
+	}
+	return nextBuild, hasRunningBuilds, nil
+}
+
+// handleComplete represents the default OnComplete handler. This Handler will
+// check which build should be run next and update the StartTimestamp field for
+// that build. That will trigger HandleBuild() to process that build immediately
+// and as a result the build is immediately executed.
+func handleComplete(lister buildclient.BuildLister, updater buildclient.BuildUpdater, build *buildapi.Build) error {
+	bcName := buildutil.ConfigNameForBuild(build)
+	if len(bcName) == 0 {
+		return fmt.Errorf("build %s/%s has no build config label set", build.Namespace, build.Name)
+	}
+	nextBuild, hasRunningBuilds, err := GetNextConfigBuild(lister, build.Namespace, bcName)
+	if err != nil {
+		return fmt.Errorf("unable to get the next build for %s/%s: %v", build.Namespace, build.Name, err)
+	}
+	if hasRunningBuilds || nextBuild == nil {
+		return nil
+	}
+	now := unversioned.Now()
+	nextBuild.Status.StartTimestamp = &now
+	if err := updater.Update(nextBuild.Namespace, nextBuild); err != nil && !errors.IsConflict(err) {
+		return err
```

Comment:
we need to be very careful we don't stall the queue by failing to update a build here.  the only thing that keeps processing going is update the build to cause the next event.

i think this is ok because the retry controller logic will retry the event that lead us down this path.